standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java

```diff
@@ -9322,9 +9334,25 @@ public void processContext(ServerContext serverContext, TTransport tTransport, T
     if (startLock != null) {
       signalOtherThreadsToStart(tServer, startLock, startCondition, startedServing);
     }
+
+    // If dynamic service discovery through ZooKeeper is enabled, add this server to the ZooKeeper.
+    if (MetastoreConf.getVar(conf, ConfVars.THRIFT_SERVICE_DISCOVERY_MODE)
+            .equalsIgnoreCase("zookeeper")) {
+      zooKeeperHelper = MetastoreConf.getZKConfig(conf);
+      String serverInstanceURI = getServerInstanceURI(port);
+      zooKeeperHelper.addServerInstanceToZooKeeper(serverInstanceURI, serverInstanceURI, null,
```

Comment:
if for some reason the server registration fails we should do a System.exit(-1) and not just leave HMS open to serving any sort of calls.

Reply:
Since we try to register before calling tServer.serve(), If an exception occurs while registering, startMetaStore() would throw it causing the process to be terminated at that point without the HMS coming up.

But I think, explicitly throwing that error with an error message logged is a better way to do it. Done.